### PR TITLE
Added split_edge method

### LINKED ISSUE #150: Add split_edge method.
Add an option that will split an edge into multiple edges. This plus `flatten_edges` are needed for certain distributed hardware optimization.

### PR DESCRIPTION
**Pull request by mcbal**
Attempt to resolve #150. First PR for this project so comments definitely welcome. Pull request also contains random typos I came across and a tiny change to the `reshape` wrapper for the numpy backend (to enable tuple support for the `shape` argument and align it with the other backends). Unit tests pass but I haven't tried the method in the context of a larger script.

### CODE REVIEW

File: tensornetwork/network_components.py [hunk 1]
```diff
@@ -1325,6 +1324,132 @@ def flatten_all_edges(nodes: Iterable[BaseNode]) -> List[Edge]:
...
+  """
+  node = edge.node1  # We are in the trace case, so this is the only node.
+  backend = node.backend
+  # Permute until edge axes to be split are at the back and reshape.
+  perm_back = [min(edge.axis1, edge.axis2)]
+  perm_back += [max(edge.axis1, edge.axis2)]
+  perm_front = set(range(len(node.edges))) - set(perm_back)
+  perm_front = sorted(perm_front)
+  node.reorder_axes(perm_front + perm_back)
+  unaffected_shape = backend.shape_tuple(node.tensor)[:len(perm_front)]
```
Comment (Thenerdstation) [inline tensornetwork/network_components.py]: I would recommend doing `backend.shape` as tensors that have an undefined shape at runtime need to have their shapes as tenors.

File: tensornetwork/network_components.py [hunk 2]
```diff
@@ -1325,6 +1324,132 @@ def flatten_all_edges(nodes: Iterable[BaseNode]) -> List[Edge]:
...
+  new_shape = unaffected_shape + shape + shape
```
Comment (Thenerdstation) [inline tensornetwork/network_components.py]: use `backend.concat`

File: tensornetwork/network_components.py [hunk 3]
```diff
@@ -1325,6 +1324,132 @@ def flatten_all_edges(nodes: Iterable[BaseNode]) -> List[Edge]:
...
+  for node in expected_nodes:
+    # Required for dangling case.
+    if node is None:
+      continue
+    # Permute until edge axes to be split are at the back and reshape.
+    perm_back = [node.edges.index(edge)]
+    perm_front = set(range(len(node.edges))) - set(perm_back)
+    perm_front = sorted(perm_front)
+    node.reorder_axes(perm_front + perm_back)
+    unaffected_shape = backend.shape_tuple(node.tensor)[:len(perm_front)]
```
Comment (Thenerdstation) [inline tensornetwork/network_components.py]: Same comment as above, use `backend.shape` instead.

File: tensornetwork/network_components.py [hunk 4]
```diff
@@ -1325,6 +1324,132 @@ def flatten_all_edges(nodes: Iterable[BaseNode]) -> List[Edge]:
...
+    new_shape = unaffected_shape + shape
```
Comment (Thenerdstation) [inline tensornetwork/network_components.py]: `backend.concat` here too.

File: tensornetwork/tests/network_test.py [hunk 5]
```diff
@@ -320,6 +320,66 @@ def test_flatten_edges_different_nodes_value_error(backend):
...
+  e2 = tn.connect(a[4], a[5])
+  external_1 = tn.connect(a[0], c[0])
+  external_2 = tn.connect(c[1], a[2])
+  shape = (2, 1, 3)
+  new_edge_names = ["New Edge 2", "New Edge 1", "New Edge 3"]
+  new_edges = tn.split_edge(e1, shape, new_edge_names)  
+  assert a.shape == (2, 4, 5, 5) + shape + shape
+  assert a.edges == [external_1, external_2, e2, e2, *new_edges, *new_edges]
+  for i, new_edge in enumerate(new_edges):
+    assert new_edge.name == new_edge_names[i]
```
Comment (Thenerdstation) [inline tensornetwork/tests/network_test.py]: You can use a `zip` here.

```python
for new_edge, new_name in zip(new_edges, new_edge_names):
  assert new_edge.name == new_name
```

File: tensornetwork/tests/network_test.py [hunk 6]
```diff
@@ -320,6 +320,66 @@ def test_flatten_edges_different_nodes_value_error(backend):
...
+  edge_b_1 = b[1] # dangling
+  shape = (2, 1, 3)
+  new_edge_names = ["New Edge 2", "New Edge 1", "New Edge 3"]
+  new_edges = tn.split_edge(e1, shape, new_edge_names)
+  assert a.shape == (3, 5) + shape
+  assert b.shape == (2, 4, 3) + shape
+  assert a.edges == [e2, edge_a_2, *new_edges]
+  assert b.edges == [edge_b_0, edge_b_1, e2, *new_edges]
+  for i, new_edge in enumerate(new_edges):
+    assert new_edge.name == new_edge_names[i]
```
Comment (Thenerdstation) [inline tensornetwork/tests/network_test.py]: Same comment, use the `zip`

File: tensornetwork/tests/network_test.py [hunk 7]
```diff
@@ -320,6 +320,66 @@ def test_flatten_edges_different_nodes_value_error(backend):
...
+  e2 = a[1]
+  e3 = a[2]
+  e4 = a[3]
+  shape = (2, 5)
+  new_edge_names = ["New Edge 2", "New Edge 5"]
+  new_edges = tn.split_edge(e2, shape, new_edge_names)
+  assert a.shape == (2, 4, 5, 2, 5)
+  assert a.edges == [e1, e3, e4, *new_edges]
+  for i, new_edge in enumerate(new_edges):
+    assert new_edge.name == new_edge_names[i]
```
Comment (Thenerdstation) [inline tensornetwork/tests/network_test.py]: `zip`

File: tensornetwork/tests/network_test.py [hunk 8]
```diff
@@ -320,6 +320,66 @@ def test_flatten_edges_different_nodes_value_error(backend):
...
+  edge_b_0 = b[0] # dangling
+  edge_b_1 = b[1] # dangling
+  shape = (2, 1, 3)
+  new_edge_names = ["New Edge 2", "New Edge 1", "New Edge 3"]
+  new_edges = tn.split_edge(e1, shape, new_edge_names)
+  assert a.shape == (3, 5) + shape
+  assert b.shape == (2, 4, 3) + shape
+  assert a.edges == [e2, edge_a_2, *new_edges]
+  assert b.edges == [edge_b_0, edge_b_1, e2, *new_edges]
+  for i, new_edge in enumerate(new_edges):
```
Comment (Thenerdstation) [inline tensornetwork/tests/network_test.py]: Can you also add checks the the `new_edge.dimesion` on all the tests?

File: tensornetwork/tests/network_test.py [hunk 9]
```diff
@@ -320,6 +320,66 @@ def test_flatten_edges_different_nodes_value_error(backend):
...
+  new_edge_names = ["New Edge 2", "New Edge 1", "New Edge 3"]
+  new_edges = tn.split_edge(e1, shape, new_edge_names)  
+  assert a.shape == (2, 4, 5, 5) + shape + shape
+  assert a.edges == [external_1, external_2, e2, e2, *new_edges, *new_edges]
+  for i, new_edge in enumerate(new_edges):
+    assert new_edge.name == new_edge_names[i]
+  tn.check_correct({a, c})
+
+
+def test_split_edges_standard(backend):
```
Comment (Thenerdstation) [inline tensornetwork/tests/network_test.py]: Can you add a test where you split a standard edge and then run `contract_between` and make sure the resulting tensor values are still correct?

Comment (mcbal) [reply tensornetwork/tests/network_test.py]: Thank you for this. This actually exposed a bug (I forgot to update the `axis_names` in `split_edge` leading to problems when calling `_remove_edges` in `contract_between`).

Comment (mcbal) [reply tensornetwork/tests/network_test.py]: And another bug: `reorder_axes` doesn't permute the axis names like `reorder_edges` does.

Comment (Thenerdstation) [reply tensornetwork/tests/network_test.py]: Good catch!
